File: experiments_realworld/graphs/problem03/implementations/suboptimal/solution.py
import sys

def main():
    data = sys.stdin.buffer.read().split()
    it = iter(data)
    
    n = int(next(it))
    q = int(next(it))
    
    next_planet = [0] * (n + 1)
    for i in range(1, n + 1):
        next_planet[i] = int(next(it))
    
    # SUBOPTIMAL: naive O(q*k) simulation instead of O(q*log k) binary lifting.
    # Walks the k teleporters one by one; for large k (up to 10^9) this is far
    # slower than the optimal, while computing the same answer.
    results = []
    for _ in range(q):
        x = int(next(it))
        k = int(next(it))
        for _step in range(k):
            x = next_planet[x]
        results.append(str(x))
    
    print('\n'.join(results))

if __name__ == "__main__":
    main()

File: experiments_realworld/graphs/problem03/implementations/suboptimal/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n, q;
    cin >> n >> q;
    
    vector<int> next(n + 1);
    for (int i = 1; i <= n; i++) {
        cin >> next[i];
    }
    
    // SUBOPTIMAL: naive O(q*k) simulation instead of O(q*log k) binary lifting.
    // Walks the k teleporters one by one; for large k (up to 10^9) this is far
    // slower than the optimal, while computing the same answer.
    for (int i = 0; i < q; i++) {
        int x, k;
        cin >> x >> k;
        for (int step = 0; step < k; step++) {
            x = next[x];
        }
        cout << x << "\n";
    }

    return 0;
}


